django/contrib/admin/options.py

```diff
@@ -1583,6 +1584,22 @@ def add_view(self, request, form_url='', extra_context=None):
     def change_view(self, request, object_id, form_url='', extra_context=None):
         return self.changeform_view(request, object_id, form_url, extra_context)
 
+    def get_edited_object_pks(self, request, prefix):
+        pk_pattern = re.compile('{prefix}-\d+-{pk_name}$'.format(prefix=prefix, pk_name=self.model._meta.pk.name))
+        return [
+            value for key, value in request.POST.items() if pk_pattern.match(key)
+        ]
+
+    def get_list_editable_queryset(self, request, prefix):
+        object_pks = self.get_edited_object_pks(request, prefix)
+        modified_objects = self.get_queryset(request).filter(pk__in=object_pks)
+        try:
+            # Force evaluate queryset to verify that pks are valid
+            list(modified_objects)
```

Comment:
I guess we could try calling the primary key's `to_python` instead of hitting the database here.

```python
def get_list_editable_queryset(self, request, prefix):
    object_pks = self.get_edited_object_pks(request, prefix)
    queryset  = self.get_queryset(request)
    validate = queryset.model._meta.pk.to_python
    try:
        for pk in object_pks:
            validate(pk)
    except ValidationError:
        # Disable optimization
        return queryset
    return queryset.filter(pk__in=object_pks)
```